File: LeetCode/Challenges/2021/January_LeetCoding_Challenge/java/24.merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> minHeap = new PriorityQueue<ListNode>((a, b) -> a.val - b.val);
        
        for (ListNode l : lists)
            if (l != null) minHeap.add(l);
        
        ListNode dummy = new ListNode();
        ListNode tail = dummy;
        while (minHeap.size() != 0) {
            tail.next = minHeap.peek();
            minHeap.poll();
            tail = tail.next;
            if (tail.next != null)
                minHeap.add(tail.next);
        }
        return dummy.next;
    }
}

File: LeetCode/Challenges/2021/January_LeetCoding_Challenge/cpp/24.merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b) { return a->val > b->val; };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> minHeap (cmp);

        for (auto l : lists)
            if (l) minHeap.push(l);

        ListNode dummy;
        ListNode *tail = &dummy;
        while (!minHeap.empty()) {
            tail->next = minHeap.top();
            minHeap.pop();
            tail = tail->next;
            if (tail->next)
                minHeap.push(tail->next);
        }
        return dummy.next;
    }
};